File: android/app/src/main/java/io/github/shiot0/shou/Wol.kt
package io.github.shiot0.shou

import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress

/**
 * Wake-on-LAN. Turning a sleeping Shou PC on from the couch is the whole point of
 * the phone being a "remote" — so each saved server can store the PC's MAC and we
 * broadcast a standard magic packet to wake it before connecting.
 *
 * A magic packet is 6×0xFF followed by the 6-byte MAC repeated 16 times, sent as a
 * UDP broadcast (so it reaches a host that has no IP yet because it's asleep).
 */
object Wol {

    private val PORTS = intArrayOf(9, 7)  // common WOL ports; send to both

    /** Returns true if a packet was sent. Accepts MACs as AA:BB:.. / AA-BB-.. / aabb.. */
    fun wake(mac: String): Boolean {
        val bytes = parseMac(mac) ?: return false
        val packet = ByteArray(6 + 16 * 6)
        for (i in 0 until 6) packet[i] = 0xFF.toByte()
        for (rep in 0 until 16) System.arraycopy(bytes, 0, packet, 6 + rep * 6, 6)

        var sent = false
        // 255.255.255.255 reaches the local subnet; also try a couple of broadcast-ish
        // targets so it works across the handful of router setups people actually have.
        val targets = listOf("255.255.255.255")
        try {
            DatagramSocket().use { sock ->
                sock.broadcast = true
                for (host in targets) {
                    val addr = InetAddress.getByName(host)
                    for (port in PORTS) {
                        try {
                            sock.send(DatagramPacket(packet, packet.size, addr, port))
                            sent = true
                        } catch (e: Exception) {
                            // keep trying the other port/target
                        }
                    }
                }
            }
        } catch (e: Exception) {
            return sent
        }
        return sent
    }

    /** 12 hex digits (with optional :/-/. separators) -> 6 bytes, or null if malformed. */
    private fun parseMac(mac: String): ByteArray? {
        val hex = mac.filter { it.isLetterOrDigit() }
        if (hex.length != 12) return null
        return try {
            ByteArray(6) { i -> hex.substring(i * 2, i * 2 + 2).toInt(16).toByte() }
        } catch (e: NumberFormatException) {
            null
        }
    }
}

File: ios/Shou/Core/Wol.swift
//  Wol.swift
//  Wake-on-LAN. Turning a sleeping Shou PC on from the couch is the whole point of
//  the phone being a "remote", so each saved server can store the PC's MAC and we
//  broadcast a standard magic packet (6×0xFF + 16× MAC) over UDP to wake it.
//
//  iOS gates UDP broadcast behind the `com.apple.developer.networking.multicast`
//  entitlement (Apple-approved per app id) — see Shou.entitlements. Ported from Wol.kt.

import Foundation
import Network

enum Wol {

    private static let ports: [UInt16] = [9, 7]   // common WOL ports; send to both

    /// Build + broadcast the magic packet. Accepts MACs as AA:BB:.. / AA-BB-.. / aabb..
    /// Returns true if at least one datagram went out.
    @discardableResult
    static func wake(_ mac: String) async -> Bool {
        guard let bytes = parseMac(mac) else { return false }
        var packet = Data(repeating: 0xFF, count: 6)
        for _ in 0..<16 { packet.append(contentsOf: bytes) }

        var sent = false
        for port in ports {
            if await sendBroadcast(packet, port: port) { sent = true }
        }
        return sent
    }

    private static func sendBroadcast(_ packet: Data, port: UInt16) async -> Bool {
        await withCheckedContinuation { (cont: CheckedContinuation<Bool, Never>) in
            let endpoint = NWEndpoint.hostPort(
                host: NWEndpoint.Host("255.255.255.255"),
                port: NWEndpoint.Port(rawValue: port)!
            )
            let conn = NWConnection(to: endpoint, using: .udp)
            var resumed = false
            func finish(_ ok: Bool) {
                guard !resumed else { return }
                resumed = true
                conn.cancel()
                cont.resume(returning: ok)
            }
            conn.stateUpdateHandler = { state in
                switch state {
                case .ready:
                    conn.send(content: packet, completion: .contentProcessed { error in
                        finish(error == nil)
                    })
                case .failed, .cancelled:
                    finish(false)
                default:
                    break
                }
            }
            conn.start(queue: .global(qos: .utility))
            // Safety net so a stuck connection can't hang the caller.
            DispatchQueue.global().asyncAfter(deadline: .now() + 2) { finish(false) }
        }
    }

    /// 12 hex digits (with optional :/-/. separators) -> 6 bytes, or nil if malformed.
    private static func parseMac(_ mac: String) -> [UInt8]? {
        let hex = mac.filter(\.isHexDigit)
        guard hex.count == 12 else { return nil }
        var out = [UInt8]()
        var idx = hex.startIndex
        for _ in 0..<6 {
            let next = hex.index(idx, offsetBy: 2)
            guard let b = UInt8(hex[idx..<next], radix: 16) else { return nil }
            out.append(b)
            idx = next
        }
        return out
    }
}
